<a href="https://colab.research.google.com/github/veniaibinder/DS-Unit-1-Sprint-2-Statistics/blob/master/DS-Unit-1-Sprint-2-Statistics/module1/LS_DS_121_t_tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [1]:
### YOUR WORK HERE
import pandas as pd
from google.colab import files
data_to_load = files.upload()

Saving imports-85.data to imports-85.data


In [2]:
column_headers = ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 
                 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length',
                 'width', 'height', 'curb-weight', 'engine-type','num-of-cylinders', 'engine-size',
                 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']

df = pd.read_csv('imports-85.data', names=column_headers)

print(df.shape)
df.head()

(205, 26)


symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

In [3]:
import numpy as np

df = df.replace({'?': np.NaN})

df.head(15)

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0           3               NaN  alfa-romero  ...       21          27  13495
1           3               NaN  alfa-romero  ...       21          27  16500
2           1               NaN  alfa-romero  ...       19          26  16500
3           2               164         audi  ...       24          30  13950
4           2               164         audi  ...       18          22  17450
5           2               NaN         audi  ...       19          25  15250
6           1               158         audi  ...       19          25  17710
7           1               NaN         audi  ...       19          25  18920
8           1               158         audi  ...       17          20  23875
9           0               NaN         audi  ...       16          22    NaN
10          2               192          bmw  ...       23          29  16430
11          0               192          bmw  ...       23          29  16925
12          0               188          bmw  ...       21          28  20970
13          0               188          bmw  ...       21          28  21105
14          1               NaN          bmw  ...       20          25  24565

[15 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and compare your sample mean estimate to that value. Can you say that they are different? 

In [0]:
### YOUR WORK HERE

In [0]:
random = df.sample(30)
random.head()

symboling normalized-losses        make  ... city-mpg highway-mpg  price
40           0                85       honda  ...       27          33  10295
183          2               122  volkswagen  ...       27          34   7975
33           1               101       honda  ...       30          34   6529
36           0                78       honda  ...       30          34   7295
11           0               192         bmw  ...       23          29  16925

[5 rows x 26 columns]

In [4]:
df['highway-mpg'].describe()

count    205.000000
mean      30.751220
std        6.886443
min       16.000000
25%       25.000000
50%       30.000000
75%       34.000000
max       54.000000
Name: highway-mpg, dtype: float64

In [0]:
random['highway-mpg'].describe()

count    30.000000
mean     29.866667
std       6.089750
min      19.000000
25%      25.000000
50%      30.000000
75%      34.000000
max      43.000000
Name: highway-mpg, dtype: float64

The population mean and the sample mean is not exactly the same but close. The sample mean is 29.9 and the population mean is 30.7. There is 0.9 point different between population and sample mean.

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [6]:
### YOUR WORK HERE
random_fuel = df.sample(10)
random_fuel['highway-mpg'].describe()

count    10.000000
mean     28.700000
std       5.229192
min      18.000000
25%      26.250000
50%      31.000000
75%      32.000000
max      33.000000
Name: highway-mpg, dtype: float64

Based on the sample of 10 cars, the average fuel efficiency is 28.7 miles per gallon on the highway.

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, what would you conclude when sampling 100 cars?

In [7]:
### YOUR WORK HERE
random_fuel_100 = df.sample(100)
random_fuel_100['highway-mpg'].describe()

count    100.000000
mean      30.370000
std        7.575419
min       16.000000
25%       24.750000
50%       29.500000
75%       34.000000
max       54.000000
Name: highway-mpg, dtype: float64

Based on the sample of 100 cars, the average fuel efficiency is 30.4 miles per gallon on the highway.

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

The salesman claim is not accurate. From both sample (10 and 100 cars), the average of fuel efficiency is not close to 35 miles per galon on the highway.
the result from both sample is slightly different because the sample size is different.

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [9]:
### YOUR WORK HERE
data_to_load = files.upload()

Saving house-votes-84.data to house-votes-84.data


In [60]:
column_headers = ['Class Name', 'handicapped-infants', 'water-project-cost-sharing', 'adoption-of-the-budget-resolution',
                  'physician-fee-freeze', 'el-salvador-aid', 'religious-groups-in-schools', 'anti-satellite-test-ban', 'aid-to-nicaraguan-contras',
                   'mx-missile', 'immigration', 'synfuels-corporation-cutback', 'education-spending',
                  'superfund-right-to-sue', 'crime', 'duty-free-exports', 'export-administration-act-south-africa']

df_house = pd.read_csv('house-votes-84.data', names=column_headers)


print(df_house.shape)
df_house.head()

(435, 17)


Class Name  ... export-administration-act-south-africa
0  republican  ...                                      y
1  republican  ...                                      ?
2    democrat  ...                                      n
3    democrat  ...                                      y
4    democrat  ...                                      y

[5 rows x 17 columns]

In [61]:
df_house = df_house.replace({'?': np.NaN})
df_house = df_house.replace({'y': 1})
df_house =df_house.replace({'n': 0})
df_house.head()

Class Name  ...  export-administration-act-south-africa
0  republican  ...                                     1.0
1  republican  ...                                     NaN
2    democrat  ...                                     0.0
3    democrat  ...                                     1.0
4    democrat  ...                                     1.0

[5 rows x 17 columns]

In [62]:
dfrepublican = df_house[df_house['Class Name']=='republican']

dfrepublican.head()

Class Name  ...  export-administration-act-south-africa
0   republican  ...                                     1.0
1   republican  ...                                     NaN
7   republican  ...                                     1.0
8   republican  ...                                     1.0
10  republican  ...                                     0.0

[5 rows x 17 columns]

In [63]:
dfrepublican.describe()

handicapped-infants  ...  export-administration-act-south-africa
count           165.000000  ...                              146.000000
mean              0.187879  ...                                0.657534
std               0.391804  ...                                0.476168
min               0.000000  ...                                0.000000
25%               0.000000  ...                                0.000000
50%               0.000000  ...                                1.000000
75%               0.000000  ...                                1.000000
max               1.000000  ...                                1.000000

[8 rows x 16 columns]

In [68]:
dfdemocrat = df_house[df_house['Class Name']=='democrat']
dfdemocrat.head()

Class Name  ...  export-administration-act-south-africa
2   democrat  ...                                     0.0
3   democrat  ...                                     1.0
4   democrat  ...                                     1.0
5   democrat  ...                                     1.0
6   democrat  ...                                     1.0

[5 rows x 17 columns]

In [69]:
dfdemocrat.describe()

handicapped-infants  ...  export-administration-act-south-africa
count           258.000000  ...                              185.000000
mean              0.604651  ...                                0.935135
std               0.489876  ...                                0.246956
min               0.000000  ...                                0.000000
25%               0.000000  ...                                1.000000
50%               1.000000  ...                                1.000000
75%               1.000000  ...                                1.000000
max               1.000000  ...                                1.000000

[8 rows x 16 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level) =**handicapped-infants**, **Adoption-of-the-budget-resolution**, **aid-to-nicaraguan-contras**, **mx-missile**, **synfuels-corporation-cutback**, **duty-free-exports**, **export-administration-act-south-africa**
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level)=**physician-fee-freeze**, **el-salvador-aid**, **religious-groups-in-schools**, **immigration**, **crime**,
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)=**water-project-cost-sharing**

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

Answer

In [107]:
### YOUR WORK HERE
from scipy import stats
#null hypothesis : the two sample means are equal. Republican support this bill at the same rate that Democrats do.
#Alternative Hypothesis: They support this bill at different rates
#Confidence Level: 99%, p < 0.01
#Based on a t-statistic of -9.2, and a p-value of 1.6e-18, we reject the null hypothesis that republicans and democrats support this bill at the same rate, 
#and suggest the alternative, that republicans and democrats show different levels of support for this bill.

stats.ttest_ind(dfrepublican['handicapped-infants'], dfdemocrat['handicapped-infants'], nan_policy = 'omit')


Ttest_indResult(statistic=-9.205264294809222, pvalue=1.613440327937243e-18)

In [91]:
print(dfrepublican['handicapped-infants'].mean())
print(dfdemocrat['handicapped-infants'].mean())

0.18787878787878787
0.6046511627906976


In [72]:
#null hypothesis : the two sample means are equal. Republican support this bill at the same rate that Democrats do.
#Alternative Hypothesis: They support this bill at different rates
#Confidence Level: 99%
#Based on a t-statistic of 0.089, and a p-value of 0.9, we are fail to reject the null hypothesis and republicans and democrats support this bill at the same rate, 

stats.ttest_ind(dfrepublican['water-project-cost-sharing'], dfdemocrat['water-project-cost-sharing'], nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

In [92]:
print(dfrepublican['water-project-cost-sharing'].mean())
print(dfdemocrat['water-project-cost-sharing'].mean())

0.5067567567567568
0.502092050209205


In [74]:
#null hypothesis : the two sample means are equal. Republican support this bill at the same rate that Democrats do.
#Alternative Hypothesis: They support this bill at different rates
#Confidence Level: 99%
#Based on a t-statistic of -23.21, and a p-value of 2.07e77, we reject the null hypothesis that republicans and democrats support this bill at the same rate, 
#and suggest the alternative, that republicans and democrats show different levels of support for this bill.
stats.ttest_ind(dfrepublican['adoption-of-the-budget-resolution'], dfdemocrat['adoption-of-the-budget-resolution'], nan_policy='omit')

Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77)

In [93]:
print(dfrepublican['adoption-of-the-budget-resolution'].mean())
print(dfdemocrat['adoption-of-the-budget-resolution'].mean())

0.13414634146341464
0.8884615384615384


In [94]:
#null hypothesis : the two sample means are equal. Republican support this bill at the same rate that Democrats do.
#Alternative Hypothesis: They support this bill at different rates
#Confidence Level: 99%
#Based on a t-statistic of 49, and a p-value of 1.99e177, we reject the null hypothesis that republicans and democrats support this bill at the same rate, 
#and suggest the alternative, that republicans and democrats show different levels of support for this bill.
stats.ttest_ind(dfrepublican['physician-fee-freeze'], dfdemocrat['physician-fee-freeze'], nan_policy='omit')

Ttest_indResult(statistic=49.36708157301406, pvalue=1.994262314074344e-177)

In [95]:
print(dfrepublican['physician-fee-freeze'].mean())
print(dfdemocrat['physician-fee-freeze'].mean())

0.9878787878787879
0.05405405405405406


In [77]:
#null hypothesis : the two sample means are equal. Republican support this bill at the same rate that Democrats do.
#Alternative Hypothesis: They support this bill at different rates
#Confidence Level: 99%
#Based on a t-statistic of 21.14, and a p-value of 5.6e-68, we reject the null hypothesis that republicans and democrats support this bill at the same rate, 
#and suggest the alternative, that republicans and democrats show different levels of support for this bill.
stats.ttest_ind(dfrepublican['el-salvador-aid'], dfdemocrat['el-salvador-aid'], nan_policy='omit')

Ttest_indResult(statistic=21.13669261173219, pvalue=5.600520111729011e-68)

In [96]:
print(dfrepublican['el-salvador-aid'].mean())
print(dfdemocrat['el-salvador-aid'].mean())

0.9515151515151515
0.21568627450980393


In [78]:
#null hypothesis : the two sample means are equal. Republican support this bill at the same rate that Democrats do.
#Alternative Hypothesis: They support this bill at different rates
#Confidence Level: 99%
#Based on a t-statistic of 9.73, and a p-value of 2.39e-20, we reject the null hypothesis that republicans and democrats support this bill at the same rate, 
#and suggest the alternative, that republicans and democrats show different levels of support for this bill.
stats.ttest_ind(dfrepublican['religious-groups-in-schools'], dfdemocrat['religious-groups-in-schools'], nan_policy='omit')

Ttest_indResult(statistic=9.737575825219457, pvalue=2.3936722520597287e-20)

In [97]:
print(dfrepublican['religious-groups-in-schools'].mean())
print(dfdemocrat['religious-groups-in-schools'].mean())

0.8975903614457831
0.47674418604651164


In [80]:
#null hypothesis : the two sample means are equal. Republican support this bill at the same rate that Democrats do.
#Alternative Hypothesis: They support this bill at different rates
#Confidence Level: 99%
#Based on a t-statistic of -18.05, and a p-value of 2.82e-54, we reject the null hypothesis that republicans and democrats support this bill at the same rate, 
#and suggest the alternative, that republicans and democrats show different levels of support for this bill.
stats.ttest_ind(dfrepublican['aid-to-nicaraguan-contras'], dfdemocrat['aid-to-nicaraguan-contras'], nan_policy='omit')

Ttest_indResult(statistic=-18.052093200819733, pvalue=2.82471841372357e-54)

In [98]:
print(dfrepublican['aid-to-nicaraguan-contras'].mean())
print(dfdemocrat['aid-to-nicaraguan-contras'].mean())

0.15286624203821655
0.8288973384030418


In [81]:
#null hypothesis : the two sample means are equal. Republican support this bill at the same rate that Democrats do.
#Alternative Hypothesis: They support this bill at different rates
#Confidence Level: 99%
#Based on a t-statistic of -16.4, and a p-value of 5.03e-47, we reject the null hypothesis that republicans and democrats support this bill at the same rate, 
#and suggest the alternative, that republicans and democrats show different levels of support for this bill.
stats.ttest_ind(dfrepublican['mx-missile'], dfdemocrat['mx-missile'], nan_policy='omit')

Ttest_indResult(statistic=-16.437503268542994, pvalue=5.03079265310811e-47)

In [99]:
print(dfrepublican['mx-missile'].mean())
print(dfdemocrat['mx-missile'].mean())

0.11515151515151516
0.7580645161290323


In [82]:
#null hypothesis : the two sample means are equal. Republican support this bill at the same rate that Democrats do.
#Alternative Hypothesis: They support this bill at different rates
#Confidence Level: 99%
#Based on a t-statistic of 1.7, and a p-value of 0.08, we reject the null hypothesis that republicans and democrats support this bill at the same rate, 
#and suggest the alternative, that republicans and democrats show different levels of support for this bill.
stats.ttest_ind(dfrepublican['immigration'], dfdemocrat['immigration'], nan_policy='omit')

Ttest_indResult(statistic=1.7359117329695164, pvalue=0.08330248490425066)

In [100]:
print(dfrepublican['immigration'].mean())
print(dfdemocrat['immigration'].mean())

0.5575757575757576
0.4714828897338403


In [83]:
#null hypothesis : the two sample means are equal. Republican support this bill at the same rate that Democrats do.
#Alternative Hypothesis: They support this bill at different rates
#Confidence Level: 99%
#Based on a t-statistic of -8.3, and a p-value of 1.5e-15, we reject the null hypothesis that republicans and democrats support this bill at the same rate, 
#and suggest the alternative, that republicans and democrats show different levels of support for this bill.
stats.ttest_ind(dfrepublican['synfuels-corporation-cutback'], dfdemocrat['synfuels-corporation-cutback'], nan_policy='omit')

Ttest_indResult(statistic=-8.293603989407588, pvalue=1.5759322301054064e-15)

In [101]:
print(dfrepublican['synfuels-corporation-cutback'].mean())
print(dfdemocrat['synfuels-corporation-cutback'].mean())

0.1320754716981132
0.5058823529411764


In [84]:
#null hypothesis : the two sample means are equal. Republican support this bill at the same rate that Democrats do.
#Alternative Hypothesis: They support this bill at different rates
#Confidence Level: 99%
#Based on a t-statistic of 20.5, and a p-value of 1.88e-64, we reject the null hypothesis that republicans and democrats support this bill at the same rate, 
#and suggest the alternative, that republicans and democrats show different levels of support for this bill.
stats.ttest_ind(dfrepublican['education-spending'], dfdemocrat['education-spending'], nan_policy='omit')

Ttest_indResult(statistic=20.500685724563073, pvalue=1.8834203990450192e-64)

In [102]:
print(dfrepublican['education-spending'].mean())
print(dfdemocrat['education-spending'].mean())

0.8709677419354839
0.14457831325301204


In [85]:
#null hypothesis : the two sample means are equal. Republican support this bill at the same rate that Democrats do.
#Alternative Hypothesis: They support this bill at different rates
#Confidence Level: 99%
#Based on a t-statistic of 13.5, and a p-value of 1.23e-34, we reject the null hypothesis that republicans and democrats support this bill at the same rate, 
#and suggest the alternative, that republicans and democrats show different levels of support for this bill.
stats.ttest_ind(dfrepublican['superfund-right-to-sue'], dfdemocrat['superfund-right-to-sue'], nan_policy='omit')

Ttest_indResult(statistic=13.51064251060933, pvalue=1.2278581709672758e-34)

In [103]:
print(dfrepublican['superfund-right-to-sue'].mean())
print(dfdemocrat['superfund-right-to-sue'].mean())

0.8607594936708861
0.2896825396825397


In [86]:
#null hypothesis : the two sample means are equal. Republican support this bill at the same rate that Democrats do.
#Alternative Hypothesis: They support this bill at different rates
#Confidence Level: 99%
#Based on a t-statistic of 16.34, and a p-value of 9.95e-47, we reject the null hypothesis that republicans and democrats support this bill at the same rate, 
#and suggest the alternative, that republicans and democrats show different levels of support for this bill.
stats.ttest_ind(dfrepublican['crime'], dfdemocrat['crime'], nan_policy='omit')

Ttest_indResult(statistic=16.342085656197696, pvalue=9.952342705606092e-47)

In [104]:
print(dfrepublican['crime'].mean())
print(dfdemocrat['crime'].mean())

0.9813664596273292
0.35019455252918286


In [87]:
#null hypothesis : the two sample means are equal. Republican support this bill at the same rate that Democrats do.
#Alternative Hypothesis: They support this bill at different rates
#Confidence Level: 99%
#Based on a t-statistic of -12.85, and a p-value of 5.99e-32, we reject the null hypothesis that republicans and democrats support this bill at the same rate, 
#and suggest the alternative, that republicans and democrats show different levels of support for this bill.
stats.ttest_ind(dfrepublican['duty-free-exports'], dfdemocrat['duty-free-exports'], nan_policy='omit')

Ttest_indResult(statistic=-12.853146132542978, pvalue=5.997697174347365e-32)

In [105]:
print(dfrepublican['duty-free-exports'].mean())
print(dfdemocrat['duty-free-exports'].mean())

0.08974358974358974
0.6374501992031872


In [88]:
#null hypothesis : the two sample means are equal. Republican support this bill at the same rate that Democrats do.
#Alternative Hypothesis: They support this bill at different rates
#Confidence Level: 99%
#Based on a t-statistic of -6.8, and a p-value of 3.6e-11, we reject the null hypothesis that republicans and democrats support this bill at the same rate, 
#and suggest the alternative, that republicans and democrats show different levels of support for this bill.
stats.ttest_ind(dfrepublican['export-administration-act-south-africa'], dfdemocrat['export-administration-act-south-africa'], nan_policy='omit')

Ttest_indResult(statistic=-6.849454815841208, pvalue=3.652674361672226e-11)

In [106]:
print(dfrepublican['export-administration-act-south-africa'].mean())
print(dfdemocrat['export-administration-act-south-africa'].mean())

0.6575342465753424
0.9351351351351351


## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [0]:
### YOUR WORK HERE





### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [0]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [0]:
### YOUR WORK HERE